django/contrib/messages/storage/cookie.py

```diff
@@ -122,39 +137,65 @@ def _store(self, messages, response, remove_oldest=True, *args, **kwargs):
         returned), and add the not_finished sentinel value to indicate as much.
         """
         unstored_messages = []
-        encoded_data = self._encode(messages)
+
+        serialized_messages = MessagePartSerializer().dumps(messages)
+        encoded_data = self._encode_parts(serialized_messages)
         if self.max_cookie_size:
             # data is going to be stored eventually by SimpleCookie, which
             # adds its own overhead, which we must account for.
             cookie = SimpleCookie()  # create outside the loop
 
-            def stored_length(val):
-                return len(cookie.value_encode(val)[1])
+            def is_too_large_for_cookie(data):
+                return data and len(cookie.value_encode(data)[1]) > self.max_cookie_size
+
+            def compute_msg(some_serialized_msg):
+                return self._encode_parts(
+                    some_serialized_msg + [self._not_finished_msg],
+                    encode_empty=True,
+                )
 
-            while encoded_data and stored_length(encoded_data) > self.max_cookie_size:
+            if is_too_large_for_cookie(encoded_data):
                 if remove_oldest:
-                    unstored_messages.append(messages.pop(0))
+                    idx = bisect_keep_right(
+                        serialized_messages,
+                        fn=lambda m: is_too_large_for_cookie(compute_msg(m)),
+                    )
+                    unstored_messages = messages[:idx]
+                    encoded_data = compute_msg(serialized_messages[idx:])
                 else:
-                    unstored_messages.insert(0, messages.pop())
-                encoded_data = self._encode(
-                    messages + [self.not_finished], encode_empty=unstored_messages
-                )
+                    idx = bisect_keep_left(
+                        serialized_messages,
+                        fn=lambda m: is_too_large_for_cookie(compute_msg(m)),
```

Comment:
I wonder if we can calculate the size via simple summing up the messages (+ delimiter)?

Reply:
Ah yes, that was a brain fart on my side, I forgot that we compress and sign. I agree that option two might improve it a bit, but I don't think it is worth it -- if we consider it worth in the future we can still add it.